src/chain.js

```diff
@@ -11,19 +11,25 @@ var map = require('./map');
  *
  * Dispatches to the `chain` method of the second argument, if present,
  * according to the [FantasyLand Chain spec](https://github.com/fantasyland/fantasy-land#chain).
+ * Treats the second argument as a FantasyLand Chain, when the the second argument is a function
+ * of the form `a -> x -> b`.
  *
  * @func
  * @memberOf R
  * @since v0.3.0
  * @category List
  * @sig Chain m => (a -> m b) -> m a -> m b
+ * @sig (x -> a) -> (a -> x -> b) -> (x -> b)
```

Comment:
> In FL, the signature is
>
> ```haskell
> chain :: Chain m => m a ~> (a -> m b) -> m b
> ```
>
> and that's what our function-branch supports.

I believe you're mistaken. Here's the [implementation][1] in [sanctuary-type-classes][2]:

> ```javascript
> //  Function$prototype$chain :: (a -> b) ~> (b -> a -> c) -> (a -> c)
> var Function$prototype$chain = function(f) {
>   var chain = this;
>   return function(x) { return f(chain(x))(x); };
> };
> ```

[`Z.chain`][3] takes its arguments in the same order as `R.map`, yet the example works as expected:

```javascript
> Z.chain(R.append, R.head)([1, 2, 3])
[1, 2, 3, 1]
```

`R` and `Z` functions needn't take their arguments in the same order as the corresponding FL methods do.

> this stuff makes my head hurt

I believe it's because the various FL implementations are scattered throughout Ramda's (large) codebase, and they're not defined as methods (so one cannot easily verify that they match the type signatures in the spec).

I'd :heart: Ramda to depend on sanctuary-type-classes so all the FL implementations live in one place. It's a small library with no run-time dependencies. Depending on it would make the Ramda codebase easier to comprehend, and may even fix some bugs which have so far gone undetected.


[1]: https://github.com/sanctuary-js/sanctuary-type-classes/blob/master/index.js#L715-L719
[2]: https://github.com/sanctuary-js/sanctuary-type-classes
[3]: https://github.com/sanctuary-js/sanctuary-type-classes#chain